### PR TITLE
Provide TARGET_SAVED and TARGET_DELETED events

### PR DESCRIPTION
**Pull request by estepper**
When building tools on top of PDE's target platform framework it's often necessary to be informed about state changes in targets. PDE already uses an IEventBroker to send "workspaceTargetChanged" events. This change is about sending two new events:

    TARGET_SAVED and TARGET_DELETED

Please note that I also added the method TargetEditor.getAdapter(), which provides callers with the ITargetHandle of the editor.

### CODE REVIEW

**Comment by laeubi** [comment]
Is this really required?

Why can't you just listen for the resource changes and see if it is a target file or not?

**Comment by estepper** [comment]
> Is this really required?

Yes.

> Why can't you just listen for the resource changes and see if it is a target file or not?

Not all target definitions are contained in IFiles. By default, actually, they're contained in normal files hidden somewhere under .metadata/.plugins/o.e.pde.core/

**Comment by laeubi** [comment]
> By default, actually, they're contained in normal files hidden somewhere under .metadata/.plugins/o.e.pde.core/

The I'm a bit confused about the use-case, if you want to capture general save (not sure about delete) I would expect only a change in the target editor **or** handled it in the target handle / target service (so we capture any save) but not on different places.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java [hunk 1]
```diff
@@ -184,6 +186,9 @@ public void doSave(IProgressMonitor monitor) {
 		fDirty = false;
 		editorDirtyStateChanged();
 		fInputHandler.setSaving(false);
+
+		TargetPlatformService service = (TargetPlatformService) TargetPlatformService.getDefault();
```
Comment (laeubi) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java:190]: Why is this needed? Should not saving the target file already trigger an event by default?

**Comment by estepper** [comment]
> The I'm a bit confused about the use-case, if you want to capture general save (not sure about delete) I would expect only a change in the target editor **or** handled it in the target handle / target service (so we capture any save) but not on different places.

That changes are needed in many different places is caused by PDE's core design. For instance, have a look at how many places are required to maintain the weird TargetPlatformHelper.fgCachedTargetDefinitionMap!

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java (hunk 1 above)
Comment (estepper) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java:190]: Only if it's the active target.

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/TargetPlatformService.java [hunk 2]
```diff
@@ -365,24 +368,30 @@ public void setWorkspaceTargetDefinition(ITargetDefinition target, boolean async
...
 			} else {
-				notifyTargetChanged(target);
+				notifyEvent(TargetEvents.TOPIC_WORKSPACE_TARGET_CHANGED, target);
 			}
 		}
 	}
 
-	static void notifyTargetChanged(ITargetDefinition target) {
-		IEclipseContext context = EclipseContextFactory.getServiceContext(PDECore.getDefault().getBundleContext());
+	public final void scheduleEvent(String topic, Object data) {
```
[the remark this answers is not in the document]
Comment (estepper) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/TargetPlatformService.java:378]: Counter question: Why have them?

[the remark this answers is not in the document]
Comment (laeubi) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/TargetPlatformService.java:378]: To hide implementation details on how the actual notifications are performed, what data is included and so on so we later can change it. Beside that it won't require an change of the internal API but just adding new API what is often less disruptive.

**Comment by estepper** [comment]
I know that EclipseCon is coming soon and is likely causing a lot of work, too. But is there a chance to get this in for M2? I have large downstream changes waiting for it in Oomph 😄

**Comment by HannesWell** [comment]
I'll try to have a look at it too, but I first have to finish my preparation for EclipseCon.
So I believe the chances to get this ready before the weekend are not very high. Sorry for that.

> I have large downstream changes waiting for it in Oomph 😄

Can't you use temporarily a fixed I-build repository in Oomph once this is done?
E.g. https://download.eclipse.org/eclipse/updates/4.34-I-builds/I20241014-0530/ won't change it's content, it will just vanish after the release.

**Comment by laeubi** [comment]
> But is there a chance to get this in for M2? I have large downstream changes waiting for it in Oomph 😄

If its related to Oomph maybe @merks can have a look.

By the way sharing a PR that demonstrate the usage of a new feature might also help to convince maintainers its a good addition :-)

At the moment at least for me it is not completely clear if/how this would be actually used. In the initial text you wrote

> When building tools on top of PDE's target platform framework it's often necessary to be informed about state changes in targets. 

In general there is only **one** target platform state and that's why there is an event if it changes the target definition to be used for that, so why is it important to know if a target file (?) was edited that is not active at all?

As mentioned above if it is about when editing a file I would then see this as a matter of the Target editor or whoever changes that target file as such a file can be e.g. edited/deleted in the file system as well.

Also this

> Please note that I also added the method TargetEditor.getAdapter(), which provides callers with the ITargetHandle of the editor.

Makes me feel you actually want some specific behavior from the target editor itself but also with this change its a bit unclear, e.g. why one can't use some other means (e.g. using the `EditorInput`) and how/when this has to be used especially as PDE itself can work without such feature.

**Comment by estepper** [comment]
In the branch "target-monitor" I implemented a new ViewPart that monitors targets and creates content snapshots which can be compared to each other in order to get a clear picture about the impact of target definition changes. You can check it out from https://github.com/eclipse-oomph/oomph/tree/target-monitor . Here's a screenshot:

image

The new code in TargetEditor.getAdapter() is needed by the "Link With Editor" functionality of the new view.

**Comment by laeubi** [comment]
@estepper is this really Omph specific or do you think it could be useful to implement such view in PDE directly e.g. does it depend on any oomph specific classes/function?

**Comment by estepper** [comment]
It uses a few Oomph-specific classes and a Nebula TableCombo widget. Apart from that I can see that at some point it could be moved to PDE. But especially at this early point I expect that additional changes will be needed to address initial problems and add smaller enhancements. And for those I'm much faster when I have it under my own control.

**Review by HannesWell — COMMENTED** [review]
Hi @estepper, sorry for the delayed reply.

I finally had a look into this PR and think that it's a suitable solution for your problem, although it's not ideal, but that's due to event notification mechanism that's already in use.
Please see my detailed remarks below.

I have a few minor clean-ups that should be applied to this PR, but since I'm currently short in time I'll add the details tomorrow.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java [hunk 3]
```diff
@@ -286,6 +291,17 @@ public void dispose() {
...
+	public <T> T getAdapter(Class<T> adapter) {
+		if (adapter.equals(ITargetHandle.class)) {
+			ITargetDefinition target = getTarget();
+			if (target != null) {
+				return adapter.cast(target.getHandle());
+			}
+		}
+
+		return super.getAdapter(adapter);
+	}
```
[the remark this answers is not in the document]
Comment (estepper) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java:304]: > Is this directly related to the main topic of this change or just something you want to change too?

Hm, I guess the latter. Above in this PR I explained it: 
The new code in TargetEditor.getAdapter() is needed by the "Link With Editor" functionality of the new view.

[the remark this answers is not in the document]
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/targetdefinition/TargetEditor.java]: Acknoledge. I oversaw that remark.

**Review by HannesWell — APPROVED** [review]
With https://github.com/eclipse-pde/eclipse.pde/pull/1466 I have made a few simplifications in the TargetPlatformService class that also simplify this PR. I have rebased this and added a second commit with suggestions for improvements, a few clean-ups and unification of the doc so that there is less duplication.
@estepper if there's no objection from your side I'll squash the suggestions into one commit and submit it this evening.

**Comment by HannesWell** [comment]
> In the branch "target-monitor" I implemented a new ViewPart that monitors targets and creates content snapshots which can be compared to each other in order to get a clear picture about the impact of target definition changes. You can check it out from https://github.com/eclipse-oomph/oomph/tree/target-monitor . Here's a screenshot:

That's very interesting.

> It uses a few Oomph-specific classes and a Nebula TableCombo widget. Apart from that I can see that at some point it could be moved to PDE. But especially at this early point I expect that additional changes will be needed to address initial problems and add smaller enhancements. And for those I'm much faster when I have it under my own control.

Oomph has multiple quite convenient tools that would (in same cases maybe a bit generalized) also fit very well into PDE. So if you are interested in moving them into the Eclipse-SDK I would be glad to discuss that.

**Comment by HannesWell** [comment]
The comparator differences are unrelated (see https://github.com/eclipse-pde/eclipse.pde/pull/1467), submitting.

Thanks again Eike for this contribution.
